src/concat.js

```diff
@@ -25,12 +25,12 @@ var _isArray = require('./internal/_isArray');
  *      R.concat([4, 5, 6], [1, 2, 3]); //=> [4, 5, 6, 1, 2, 3]
  *      R.concat('ABC', 'DEF'); // 'ABCDEF'
  */
-module.exports = _curry2(function(set1, set2) {
-  if (_isArray(set2)) {
-    return _concat(set1, set2);
-  } else if (_hasMethod('concat', set1)) {
-    return set1.concat(set2);
+module.exports = _curry2(function(a, b) {
+  if (_isArray(a) && _isArray(b)) {
+    return _concat(a, b);
+  } else if (_hasMethod('concat', a)) {
```

Comment:
If the argument is that we can't detect all invalid applications, so we shouldn't throw a descriptive exception even in cases where we *could*, we should remove the `_hasMethod('concat', a)` check. It's inconsistent with the argument.

I'm very much in favour of descriptive error messages, so my stance is that we should take reasonable steps to ensure that `a` and `b` support `concat` and are of the same type. Throwing sooner rather than later leads to shallower stack traces and better debugging experiences.